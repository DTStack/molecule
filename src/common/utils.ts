export function searchById(id) {
    return (item) => item.id === id;
}

/**
 * Clone a new object by an object instance
 * @param origin Original object instance
 */
export function cloneInstance<T>(origin: T) {
    try {
        const prototypes = Object.getPrototypeOf(origin) || {};
        Object.keys(prototypes).forEach((prop) => {
            if (typeof prototypes[prop] === 'function') {
                prototypes[prop].bind(origin);
            }
        });
        return {
            ...origin,
            ...prototypes,
        };
    } catch (e) {
        console.error('Function cloneInstance error:', e);
    }
}

/**
 * Merge multiple functions to one function
 * @param funcs
 */
export function mergeFunctions(...funcs) {
    return function (...args) {
        funcs.filter((fn) => !!fn).forEach((fn) => fn?.(...args));
    };
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function mergeObjects<T>(source: object, target: object): T {
    for (const prop in target) {
        if (typeof source[prop] === 'object' && target[prop]) {
            source[prop] = mergeObjects(source[prop], target[prop]);
        } else {
            source[prop] = target[prop];
        }
    }
    return source as any;
}

/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: object): object {
    const flatted = {};
    const recurse = (parent: string, target: object) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: object): object {
    let normalized = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = mergeObjects(normalized, root);
        }
    }
    return normalized;
}

/**
 * Determine if a color is light or dark.
 * @param color HEX or RGB
 */
export function colorLightOrDark(color: string) {
    // Variables for red, green, blue values
    let r: number;
    let g: number;
    let b: number;

    // Check the format of the color, HEX or RGB?
    if (color.match(/^rgb/)) {
        // If RGB --> store the red, green, blue values in separate variables
        const matchArray =
            color.match(
                /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
            ) || [];
        r = +matchArray[1];
        g = +matchArray[2];
        b = +matchArray[3];
    } else {
        // If hex --> Convert it to RGB
        let rgbNum = +('0x' + color.slice(1, 7));
        if (color.length < 5) {
            rgbNum = +('0x' + color.slice(1).replace(/./g, '$&$&').slice(0, 6));
        }
        r = rgbNum >> 16;
        g = (rgbNum >> 8) & 255;
        b = rgbNum & 255;
    }

    // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
    const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b));

    // Using the HSP value, determine whether the color is light or dark
    if (hsp > 127.5) {
        return 'light';
    }
    return 'dark';
}
