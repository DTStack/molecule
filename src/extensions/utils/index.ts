import type { IMenuItemProps, IterableItem } from 'mo/types';

export function createMenuDuplicate(item: IMenuItemProps): IMenuItemProps {
    return {
        ...item,
        id: `${item.id}_duplicate`,
        name: `隐藏 “${item.name}”`,
        icon: undefined,
        symbolic: item.id,
    };
}

export function createContextMenu(items: IterableItem[]): IMenuItemProps[] {
    return items.map((item) => ({
        id: item.id,
        name: item.name,
        disabled: item.disabled,
        icon: item.hidden ? undefined : 'check',
    }));
}

export function isInputEle() {
    return document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
}

/**
 * Judge current active element is monaco-editor's textarea or native element
 */
export function isNativeWidget() {
    const isEditableEle =
        document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
    if (!isEditableEle) return true;
    return !document.activeElement?.classList.contains('inputarea');
}
