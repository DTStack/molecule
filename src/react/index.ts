export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
