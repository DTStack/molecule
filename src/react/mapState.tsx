import React, { Component } from 'react';
import Logger from 'mo/common/logger';

/**
 * Mapping the state to the component
 * TODO support mapping service method to the component.
 * @param WrappedComponent The component will be wrapped
 * @param state The state you want to injected, notice the state must be an observable object
 * @param subscribes The events of your subscribe, it used to trigger the state re render
 */
export function mapState<S, T>(
    WrappedComponent: React.ComponentType<S>,
    state: S,
    actions?: T
) {
    return class StateProvider extends Component {
        state: { lastUpdated: number };
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            // There is no declare state parameter as IObservable type, so must convert to any type.
            if ((state as any).subscribe) {
                (state as any).subscribe(this.onChange);
            } else {
                Logger.error(
                    'The state parameter of mapState must be an observable object.'
                );
            }
        }
        /**
         * TODO: Performance optimize, now whatever any properties changed in target,
         * there always be trigger the onChange event, so need a compare operation.
         * @param nextState changed data
         */
        onChange(nextState: S) {
            Logger.info(nextState, state);
            this.setState({
                lastUpdated: Date.now(),
            });
        }

        render() {
            return (
                <WrappedComponent
                    {...this.state}
                    {...state}
                    {...this.props}
                    {...actions}
                />
            );
        }
    };
}
