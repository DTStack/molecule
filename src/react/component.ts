import { cloneDeep } from 'lodash';
import { EventEmitter, GlobalEvent } from 'mo/common/event';

export enum ComponentEvents {
    Update = 'Component.Update',
}

export interface IComponent<S = any> {
    /**
     * Set the Component state
     * @param values The next values of state
     * @param callback calling after setState
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Trigger the Component update event
     * @param nextState
     */
    render(nextState?: S): void;
    /**
     * Listen to the Component state update event
     * @param listener
     */
    onUpdateState(listener: (prevState: S, nextState: S) => void): void;
    /**
     * Remove the Component update event listening, default is remove all,
     * also you can remove one by pass the listener
     */
    removeOnUpdateState(listener?: Function): void;
    /**
     * Force to update the Component
     */
    forceUpdate(): void;
    /**
     * Get the Component state
     */
    getState(): S;
}

export abstract class Component<S = any>
    extends GlobalEvent
    implements IComponent<S>
{
    protected abstract state: S;
    private _event: EventEmitter;

    constructor() {
        super();
        this._event = new EventEmitter();
    }

    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    public setState(
        values: Partial<S>,
        callback?: (prevState: S, nextState: S) => void
    ) {
        const nextState = Object.assign(this.state, values);
        this.render(nextState);
        callback?.(this.state, nextState);
    }

    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    public render(nextState?: S) {
        this._event.emit(ComponentEvents.Update, this.state, nextState);
    }

    public onUpdateState(listener: (prevState: S, nextState: S) => void) {
        this._event.subscribe(ComponentEvents.Update, listener);
    }

    public removeOnUpdateState(listener?: Function): void {
        this._event.unsubscribe(ComponentEvents.Update, listener);
    }

    public forceUpdate() {
        this.setState(cloneDeep(this.state));
    }

    public getState(): S {
        return this.state;
    }
}
