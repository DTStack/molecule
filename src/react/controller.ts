import { GlobalEvent } from 'mo/common/event';

export abstract class Controller extends GlobalEvent {
    public abstract initView(): void;
}
