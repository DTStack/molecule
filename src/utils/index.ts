import { Children, cloneElement, isValidElement } from 'react';
import { merge, omitBy } from 'lodash-es';
import type { editor } from 'mo/monaco';
import type { Arraylize, IColorTheme, IconType, IMenuItemProps, IterableItem, RecordWithId, UniqueId } from 'mo/types';

export function searchById<T extends RecordWithId<Record<string, any>>>(id?: UniqueId) {
    return (item: T) => item.id === id;
}

type sortIndexRequired = { sortIndex?: number; [key: string]: any };
/**
 * The smaller the sort number is, the more front the order is
 */
export function sortByIndex(a: sortIndexRequired, b: sortIndexRequired) {
    const prev = a.sortIndex || 0;
    const next = b.sortIndex || 0;
    return prev - next;
}

/**
 * Clone react children props
 * @param children React.ReactNode
 * @param props Parent props
 */
export function cloneReactChildren<P>(children: React.ReactNode, props: P): React.ReactNode {
    return Children.map(children, (child) => {
        if (isValidElement(child)) {
            return cloneElement(child, props);
        }
        return child;
    });
}

/**
 * Make an object to be obejct array
 */
export function arraylize<T>(data: Arraylize<T>) {
    return Array.isArray(data) ? data : [data];
}

/**
 * Classify arr into two parts
 */
export function classify<T>(arr: T[], predicate: (value: T, index: number) => boolean) {
    return classifyBy(arr, (...args) => Number(!predicate(...args)));
}

/**
 * Classify arr into multiply parts
 */
export function classifyBy<T>(arr: T[], predicate: (value: T, index: number) => number) {
    return arr.reduce<T[][]>((acc, cur, index) => {
        const idx = predicate(cur, index);
        acc[idx] ??= [];
        acc[idx].push(cur);
        return acc;
    }, []);
}

export function get<T extends RecordWithId<{ children?: T[] }>>(obj: Arraylize<T>, keyPath: string[]) {
    let root = { children: arraylize(obj) } as T;
    const stack = [...keyPath];
    while (stack.length) {
        const item = stack.shift();
        const next = root.children?.find((i) => i.id === item);
        if (!next) break;
        root = next;
    }
    return root as T;
}

/**
 * get the next icon
 * @param prev previous icon
 * @param checked
 */
export function toggleNextIcon(prev?: IconType, checked?: boolean) {
    if (typeof checked === 'boolean') {
        return checked === true ? 'check' : undefined;
    }
    if (typeof prev === 'string' || typeof prev === 'undefined') {
        return prev === 'check' ? undefined : 'check';
    }
    // JSX.Element doesn't change icon status
    return prev as IconType;
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function concatMenu(...args: IMenuItemProps[][]) {
    const nonEmptyArgs = args.filter((i) => Array.isArray(i) && i.length);
    if (!nonEmptyArgs.length) return [];
    return nonEmptyArgs.reduce((acc, cur) => {
        return acc.concat({ id: randomId(), type: 'divider' }, cur);
    });
}

/**
 * Converts an object of color values into CSS variables.
 *
 * @param {Record<string, string>} colors - The object containing color values.
 * @return {string} The CSS variables string.
 */
export function convertToCSSVars(colors: Record<string, string | null>): string {
    return `
        :root {
            ${Object.keys(colors)
                .map((id) => {
                    const color = colors[id];
                    if (!color) return '';
                    const colorName = id.replace('.', '-');
                    return `--${colorName}: ${color};`;
                })
                .filter(Boolean)
                .join('\n')}
        }
    `;
}

/**
 * Converts a token to a token theme rule.
 *
 * @param {IColorTheme['tokenColors']} token - The token to be converted.
 * @return {editor.ITokenThemeRule[]} The converted token theme rule.
 */
export function convertToToken(token?: IColorTheme['tokenColors']) {
    return token?.reduce<editor.ITokenThemeRule[]>((acc, cur) => {
        if (Array.isArray(cur.scope)) {
            cur.scope.forEach((s) => () => {
                if (!s) return;
                const target = acc.find((r) => r.token === s);
                if (target) {
                    Object.assign(target, {
                        ...cur.settings,
                    });
                } else {
                    acc.push({
                        token: s,
                        ...cur.settings,
                    });
                }
            });
        }
        return acc;
    }, []);
}

export function colorsToString(colors: Record<string, string | null>) {
    return omitBy(colors, (value) => !value) as Record<string, string>;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: Record<string, any>): Record<string, any> {
    let normalized: Record<string, any> = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root: Record<string, any> = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = merge(normalized, root);
        }
    }
    return normalized;
}
/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: Record<string, any>): Record<string, any> {
    const flatted: Record<string, any> = {};
    const recurse = (parent: string, target: Record<string, any>) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

export function getPrevOrNext<T>(arr: T[], idx: number) {
    return arr[idx - 1] ?? arr[idx + 1];
}

export function isElementInParentView(ele: HTMLElement, parent: HTMLElement) {
    const elRect = ele.getBoundingClientRect();
    const parentRect = parent.getBoundingClientRect();
    const left = elRect.left >= parentRect.left;
    const right = elRect.left + elRect.width <= parentRect.left + parentRect.width;
    const top = elRect.top >= parentRect.top;
    const bottom = elRect.top + elRect.height <= parentRect.top + parentRect.height;
    const overlay = left && right && top && bottom;
    let isWhichSide: 'right' | 'left' | 'top' | 'bottom' | undefined;
    if (!overlay) {
        if (!left) isWhichSide = 'left';
        else if (!right) isWhichSide = 'right';
        else if (!top) isWhichSide = 'top';
        else if (!bottom) isWhichSide = 'bottom';
    }
    return [overlay, isWhichSide] as const;
}

export function getNameForTitle(name: IterableItem['name']) {
    return typeof name === 'string' ? name : undefined;
}
