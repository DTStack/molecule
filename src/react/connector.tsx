import 'reflect-metadata';
import { container } from 'tsyringe';
import React, { Component } from 'react';
import Logger from 'mo/common/logger';
import { IComponent } from './component';
import { Controller } from './controller';

export type ServiceObject = {
    [index: string]: IComponent;
};

export type ControllerObject = {
    [index: string]: Controller;
};

export function connect<T = any>(
    Service: IComponent | ServiceObject,
    View: React.ComponentType<any>,
    Controller?: Controller | ControllerObject
): React.ComponentType<T> {
    return class Connector extends Component<T, any> {
        state: { lastUpdated: number };
        private _isMounted = false;
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            this._isMounted = true;
            this.handleService((service) => {
                service.onUpdateState(this.onChange);
            });
        }

        componentWillUnmount() {
            this._isMounted = false;
            this.handleService((service) => {
                service.removeOnUpdateState(this.onChange);
            });
        }

        onChange(prevState, nextState) {
            Logger.info(prevState, nextState, (container as any)._registry);
            this.update();
        }

        update = () => {
            if (this._isMounted) {
                this.setState({
                    lastUpdated: Date.now(),
                });
            }
        };

        getServiceState() {
            const target = {};
            this.handleService((service, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: { ...service.getState() },
                    });
                } else {
                    Object.assign(target, { ...service.getState() });
                }
            });
            return target;
        }

        handleService(callback: (service: IComponent, prop?: string) => void) {
            if (this.isValidService(Service as IComponent)) {
                callback(Service as IComponent);
            } else {
                for (const name in Service) {
                    if (name) {
                        const service: IComponent = Service[name];
                        if (this.isValidService(service)) {
                            callback(service, name);
                        }
                    }
                }
            }
        }

        getControllerProperties() {
            const target = {};
            if (!Controller) {
                return target;
            }

            this.handleController((controller: Controller, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: this.getObjectPublicProperties(controller),
                    });
                } else {
                    Object.assign(
                        target,
                        this.getObjectPublicProperties(controller)
                    );
                }
            });
            return target;
        }

        handleController(
            callback: (controller: Controller, prop?: string) => void
        ) {
            if (this.isValidController(Controller as Controller)) {
                callback(Controller as Controller);
            } else {
                for (const name in Controller) {
                    if (name) {
                        const controller: Controller = Controller[name];
                        if (this.isValidController(controller)) {
                            callback(controller, name);
                        }
                    }
                }
            }
        }

        render() {
            return (
                <View
                    {...this.state}
                    {...this.props}
                    {...this.getServiceState()}
                    {...this.getControllerProperties()}
                />
            );
        }

        private isValidService(service: IComponent) {
            return typeof service.onUpdateState === 'function';
        }

        private isValidController(controller: Controller) {
            return typeof controller.initView === 'function';
        }

        private getObjectPublicProperties(obj: Controller) {
            const keys = Object.keys(obj);
            const result = {};
            keys.forEach((key) => {
                // Filter the Service, Controller instances and private properties which start with '_'
                // TODO There need a better way to filter the private properties of Object,
                // maybe we can make use of the # identifier in future.
                if (
                    !key.endsWith('Service') &&
                    !key.endsWith('Controller') &&
                    !key.startsWith('_')
                ) {
                    result[key] = obj[key];
                }
            });
            return result;
        }
    };
}
