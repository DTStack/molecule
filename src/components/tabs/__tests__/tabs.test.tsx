import React from 'react';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Tabs, tabsClassName } from '..';
import {
    ITabProps,
    tabItemActiveClassName,
    tabItemClassName,
    tabItemExtraClassName,
} from '../tab';
import { dragToTargetNode } from '@test/utils';

const mockData: ITabProps[] = [
    {
        id: '1',
        name: 'test-title-1',
        renderPane: <div data-testid="test1">test1</div>,
        closable: true,
    },
    {
        id: '2',
        name: 'test-title-2',
        renderPane: <div data-testid="test2">test2</div>,
        closable: true,
    },
];

afterEach(cleanup);

function mockClientOffset(boundingRectSize: number, size: number) {
    // @ts-ignore
    HTMLDivElement.prototype.getBoundingClientRect = jest.fn(() => ({
        right: boundingRectSize,
        left: boundingRectSize,
    }));

    // @ts-ignore
    Event.prototype.clientX = size;
    // @ts-ignore
    Event.prototype.clientY = size;
}

describe('The Tabs Components', () => {
    test('The snapshot', () => {
        const component = renderer.create(<Tabs data={mockData} />);
        const tabs = component.toJSON();
        expect(tabs).toMatchSnapshot();
    });

    test('Should render nothing', () => {
        const { container } = render(<Tabs />);
        const header = container.querySelector('.mo-tabs__header');

        expect(header?.childElementCount).toBe(0);
    });

    test('Default type is line', () => {
        const { container } = render(
            <Tabs data={mockData} className="test-tabs" />
        );

        const body = container.querySelector(`.${tabsClassName}`)!;
        expect(body.classList).toContain('mo-tabs--line');
        expect(body.classList).toContain('test-tabs');
    });

    test('Should render card type', () => {
        const { container } = render(<Tabs type="card" data={mockData} />);

        const body = container.querySelector(`.${tabsClassName}`)!;
        expect(body.classList).toContain('mo-tabs--card');
    });

    test('Should active specific tab', () => {
        const { getByText } = render(<Tabs activeTab="1" data={mockData} />);
        const testHeader = getByText('test-title-1');
        expect(testHeader?.classList).toContain(tabItemActiveClassName);
    });

    test('Should trigger events', async () => {
        const mockSelectTabFn = jest.fn();
        const mockContextFn = jest.fn();
        const mockCloseFn = jest.fn();
        const mockMoveFn = jest.fn();
        const { getByText } = render(
            <Tabs
                activeTab="1"
                data={mockData}
                onSelectTab={mockSelectTabFn}
                onContextMenu={mockContextFn}
                onCloseTab={mockCloseFn}
                onMoveTab={mockMoveFn}
            />
        );

        const testHeader = getByText('test-title-1');
        fireEvent.click(testHeader);

        expect(mockSelectTabFn).toBeCalled();
        expect(mockSelectTabFn.mock.calls[0][0]).toBe('1');

        fireEvent.contextMenu(testHeader);
        expect(mockContextFn).toBeCalled();

        fireEvent.mouseOver(
            testHeader.querySelector(`.${tabItemExtraClassName}`)!
        );
        await waitFor(() => {
            fireEvent.click(testHeader.querySelector('.codicon-close')!);
        });

        expect(mockCloseFn).toBeCalled();
        expect(mockCloseFn.mock.calls[0][0]).toBe('1');

        dragToTargetNode(getByText('test-title-2'), getByText('test-title-1'));
        expect(mockMoveFn).toBeCalled();
        expect(mockMoveFn.mock.calls[0][0]).toEqual([
            {
                id: '2',
                name: 'test-title-2',
                renderPane: <div data-testid="test2">test2</div>,
                closable: true,
            },
            {
                id: '1',
                name: 'test-title-1',
                renderPane: <div data-testid="test1">test1</div>,
                closable: true,
            },
        ]);
    });

    test('Should NOT onMove', () => {
        const originalBoundingClientRect =
            HTMLDivElement.prototype.getBoundingClientRect;
        //@ts-ignore
        const originalClientX = Event.prototype.clientX;
        //@ts-ignore
        const originalClientY = Event.prototype.clientY;

        const mockFn = jest.fn();
        const { container } = render(
            <Tabs activeTab="1" data={mockData} onMoveTab={mockFn} />
        );

        const tabs = container.querySelectorAll<HTMLDivElement>(
            `.${tabItemClassName}`
        )!;

        // drag up
        mockClientOffset(5, 10);
        dragToTargetNode(tabs[1], tabs[0]);
        expect(mockFn).not.toBeCalled();

        // drag down
        mockClientOffset(20, 10);
        dragToTargetNode(tabs[0], tabs[1]);
        expect(mockFn).not.toBeCalled();

        // drag to itself
        mockClientOffset(20, 10);
        dragToTargetNode(tabs[0], tabs[0]);
        expect(mockFn).not.toBeCalled();

        // reset mock function
        HTMLDivElement.prototype.getBoundingClientRect =
            originalBoundingClientRect;
        // @ts-ignore
        Event.prototype.clientX = originalClientX;
        // @ts-ignore
        Event.prototype.clientY = originalClientY;
    });
});
