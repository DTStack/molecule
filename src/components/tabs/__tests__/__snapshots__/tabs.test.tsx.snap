// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Tabs Components The snapshot 1`] = `
<div
  className="mo-tabs mo-tabs--line"
>
  <div
    className="mo-tabs__header"
  >
    <div
      className="mo-tab__item"
      onClick={[Function]}
      onContextMenu={[Function]}
    >
      test-title-1
      <div
        className="mo-tab__item__extra"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        role="button"
      >
        <span
          className="codicon codicon-close"
        />
      </div>
    </div>
    <div
      className="mo-tab__item"
      onClick={[Function]}
      onContextMenu={[Function]}
    >
      test-title-2
      <div
        className="mo-tab__item__extra"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        role="button"
      >
        <span
          className="codicon codicon-close"
        />
      </div>
    </div>
  </div>
</div>
`;
